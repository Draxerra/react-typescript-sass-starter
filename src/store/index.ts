import { configureStore, MiddlewareArray } from "@reduxjs/toolkit";
import { useDispatch, useSelector, TypedUseSelectorHook } from "react-redux";

import counter from "./counter";

const store = configureStore({
  reducer: {
    counter,
  },
  middleware: new MiddlewareArray(),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch = (): AppDispatch => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export default store;
