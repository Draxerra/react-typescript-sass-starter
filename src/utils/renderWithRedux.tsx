import { FC } from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";

type RenderParams = Parameters<typeof render>;
type Render = ReturnType<typeof render>;
type StoreParams = Parameters<typeof configureStore>;
type Store = ReturnType<typeof configureStore>;

interface RenderWithReduxOptions {
  preloadedState?: StoreParams[0]["preloadedState"];
  reducers: StoreParams[0]["reducer"];
  renderOptions?: RenderParams[1];
  store?: Store;
}
const renderWithRedux: (
  ui: RenderParams[0],
  opts: RenderWithReduxOptions
) => Render = (
  ui,
  {
    preloadedState = {},
    reducers,
    store = configureStore({ reducer: reducers, preloadedState }),
    ...renderOptions
  }
) => {
  const Wrapper: FC = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

export * from "@testing-library/react";
export { renderWithRedux };
