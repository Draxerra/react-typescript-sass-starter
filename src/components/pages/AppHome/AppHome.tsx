import { FC } from "react";
import classNames from "classnames";

import { useAppDispatch, useAppSelector } from "~/store";
import { decrement, increment } from "~/store/counter";

import AppCounter from "~/components/molecules/AppCounter";

type AppHomeProps = React.HTMLAttributes<HTMLDivElement>;
const AppHome: FC<AppHomeProps> = ({ className, ...props }) => {
  const count = useAppSelector((state) => state.counter.value);
  const dispatch = useAppDispatch();

  return (
    <div {...props} className={classNames(className, "app-home")}>
      <AppCounter
        decrement={() => dispatch(decrement())}
        increment={() => dispatch(increment())}
      >
        {count}
      </AppCounter>
    </div>
  );
};

export default AppHome;
