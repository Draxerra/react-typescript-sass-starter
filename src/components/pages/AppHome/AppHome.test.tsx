import AppHome from "./AppHome";
import { renderWithRedux } from "~/utils/renderWithRedux";
import userEvent from "@testing-library/user-event";

import counter from "~/store/counter";
import * as counterActions from "~/store/counter";

describe("App Home", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("should render", () => {
    const home = renderWithRedux(<AppHome />, {
      reducers: { counter },
    }).asFragment();
    expect(home).toMatchSnapshot();
  });

  it("should display count from store", () => {
    const { getByText } = renderWithRedux(<AppHome />, {
      reducers: { counter },
      preloadedState: { counter: { value: 500 } },
    });
    expect(getByText("Current Count: 500")).toBeTruthy();
  });

  it("should fire increment action", () => {
    const { getByText } = renderWithRedux(<AppHome />, {
      reducers: { counter },
    });
    const spy = jest.spyOn(counterActions, "increment");
    const incrementBtn = getByText("Increment");
    userEvent.click(incrementBtn);
    expect(spy).toHaveBeenCalled();
  });

  it("should increment on click", () => {
    const { getByText } = renderWithRedux(<AppHome />, {
      reducers: { counter },
    });
    const incrementBtn = getByText("Increment");
    userEvent.click(incrementBtn);
    expect(getByText("Current Count: 1")).toBeTruthy();
  });

  it("should fire decrement action", () => {
    const { getByText } = renderWithRedux(<AppHome />, {
      reducers: { counter },
    });
    const spy = jest.spyOn(counterActions, "decrement");
    const decrementBtn = getByText("Decrement");
    userEvent.click(decrementBtn);
    expect(spy).toHaveBeenCalled();
  });

  it("should decrement on click", () => {
    const { getByText } = renderWithRedux(<AppHome />, {
      reducers: { counter },
    });
    const decrementBtn = getByText("Decrement");
    userEvent.click(decrementBtn);
    expect(getByText("Current Count: -1")).toBeTruthy();
  });
});
