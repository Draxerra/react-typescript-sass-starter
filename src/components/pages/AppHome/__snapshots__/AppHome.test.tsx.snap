// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Home should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="app-home"
  >
    <div
      class="app-counter"
    >
      <span
        class="app-counter__text"
      >
        Current Count: 0
      </span>
      <button
        aria-label="Increment value"
        class="counter-increment app-button"
      >
        Increment
      </button>
      <button
        aria-label="Decrement value"
        class="counter-decrement app-button"
      >
        Decrement
      </button>
    </div>
  </div>
</DocumentFragment>
`;
